package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,307 @@
+                 start=None, stop=None, step=None, epsilon=None,
+                 weights_ker=None):
+        """
+        Parameters
+        -------------
+        u : trajectory `~MDAnalysis.core.AtomGroup.Universe`
+            The MD Trajectory for dimension reduction, remember that
+            computational cost of eigenvalue decomposition
+            scales at O(N^3) where N is the number of frames.
+            Cost can be reduced by increasing step interval or specifying a
+            start and stop.
+        select: str, optional
+            Any valid selection string for
+            :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.select_atoms`
+            This selection of atoms is used to calculate the RMSD between
+            different frames. Water should be excluded.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and
+            symmetric, Default: ``None`` sets metric to rms.rmsd()
+        cutoff : float, optional
+            Specify a given cutoff for metric values to be considered equal
+        weights : array, optional
+            Weights to be given to coordinates for metric calculation
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        epsilon : function, optional
+            Specifies the method used for the choice of scale parameter in the
+            diffusion map. More information in [1] and [2]. With 'average'
+            the average of the RMSD to the k-th greatest value is used.
+        weights_ker: list, optional
+            The list has to have the same length as the trajectory.
+            With 'None' the weight of each frame of the trajectory will be the
+            same.
+        """
+        self._u = u
+        traj = self._u.trajectory
+        self.atoms = self._u.select_atoms(select)
+
+        if metric is None:
+            self._metric = rms.rmsd
+        else:
+            self._metric = metric
+        self._cutoff = cutoff
+        self._weights = weights
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+        self._epsilon = epsilon
+
+        if weights_ker is None:
+            # weights do not apply to metric
+            self._weights_ker = np.ones((self.nframes,))
+        else:
+            if weights.shape[0] != self.nframes:
+                raise ValueError("The weight should have the same length as "
+                                 'the trajectory')
+            else:
+                # weights are constructed as relative to the mean
+                self._weights_ker = (np.asarray(weights, dtype=np.float64) /
+                                     np.mean(weights))
+
+    def _conclude(self):
```

Comment:
Thanks for all this valuable input @kain88-de! Perhaps you're right and I went the wrong way. I want to allow the function to be tuneable for custom purposes such that the way epsilon is calculated is left up to the user. In addition, it would be nice if someone working in a scripting environment could just write a function in there script and pass it as an argument. Given some explicit instructions, a user passes epsilon=function_written_for_choice_of_epsilon and the arguments necessary to execute the function. (Would this be in the form of **kwargs?, I am not entirely sure of the formalism). Of course, many epsilon functions would take the distance matrix as a parameter. Hopefully that makes sense and indicates why I think this wouldn't require recalculating the complete distance matrix.